Sources/RDeviceIdentifier/RDeviceIdentifier.swift

```diff
@@ -0,0 +1,283 @@
+    ///    do things differently and the only way to get a consistent behavior across OS versions
+    ///    with that method is to use a 3-component bundle ID, e.g. `com.rakuten.ichiba`.
+    ///
+    /// - Warning: Applications built with different application identifier prefixes/bundle seed identifiers, i.e.
+    ///          different provisioning profiles, will not produce the same device identifier.
+    ///
+    /// - Returns: A string uniquely identifying the device the application is currently running on.
+    ///         If the keychain is not available (i.e. the device is locked) and no value has been
+    ///         retrieved yet, `nil` is returned and the developer should try again when the device
+    ///         is unlocked.
+    ///
+    /// - Precondition: NSObjectInaccessibleException if the application's entitlements do not include the
+    ///            access group required to access the device identifier. See @ref device-information-keychain-setup "Setting up the keychain"
+    ///            for more information.
+    /// - Precondition: NSInternalInconsistencyException if the application is misconfigured and the first
+    ///            access group does not match the application's bundle identifier. See @ref device-information-keychain-setup "Setting up the keychain"
+    ///            for more information.
+    @objc public class var uniqueDeviceIdentifier: String? {
+        // If we already have a value, return it.
+        if let _uniqueDeviceIdentifier = _uniqueDeviceIdentifier {
+            return _uniqueDeviceIdentifier
+        }
+
+        // First, try to grab the application identifier prefix (=bundle seed it)
+        // and build the access group from it.
+        var accessGroup: String?
+        do {
+            if let result = try keychain.addIfNonExistent(),
+               let defaultAccessGroup = result[String(kSecAttrAccessGroup)] as? String,
+               let teamName = defaultAccessGroup.components(separatedBy: ".").first {
+                accessGroup = "\(teamName).\(keychainAccessGroup)"
+
+                // While we're at it, why not check developers didn't do the unthinkable?
+                checkKeychainAccessGroupOrder(defaultAccessGroup)
+            }
+        } catch {
+            // Keychain is not available
+            return nil
+        }
+
+        // Try to clean things up
+        keychain.clear()
+
+        // Try to find the device identifier in the keychain.
+        // Here we always have a bundle seed id.
+        do {
+            if let accessGroup = accessGroup, let uniqueDeviceIdData = try keychain.search(for: accessGroup) {
+                // Device id found!
+                _uniqueDeviceIdentifier = uniqueDeviceIdData.hexString
+                return _uniqueDeviceIdentifier
+            }
+        } catch {
+            // Keychain problem
+            return nil
```

Comment:
We throw errors but handle them silently by returning nil? Maybe we can at least log them (using RLogger maybe?)

Reply:
The original Obj-C version was doing something similar (catching it's own internally thrown `NSException` then returning `nil`).

Reply:
Swift 5.5 allows you to throw in computed properties now. Wonder if that would work bridged into Obj-C for Obj-C apps